tools/speedy_sync/src/main.rs

```diff
@@ -0,0 +1,272 @@
+
+    let checkpoint = SpeedyCheckpoint {
+        prev_epoch,
+        current_epoch,
+        next_epoch,
+        block,
+        prev_block,
+        final_block,
+        first_block,
+    };
+
+    let serialized = serde_json::to_string(&checkpoint).unwrap();
+
+    fs::write(Path::new(&create_cmd.destination_dir).join("snapshot.json"), serialized)
+        .expect("Failed writing to destination file");
+
+    fs::write(
+        Path::new(&create_cmd.destination_dir).join("snapshot.borsh"),
+        checkpoint.try_to_vec().unwrap(),
+    )
+    .expect("Failed writing to destination file");
+
+    fs::copy(
+        Path::new(&create_cmd.home).join("genesis.json"),
+        Path::new(&create_cmd.destination_dir).join("genesis.json"),
+    )
+    .unwrap();
+    fs::copy(
+        Path::new(&create_cmd.home).join("config.json"),
+        Path::new(&create_cmd.destination_dir).join("config.json"),
+    )
+    .unwrap();
+}
+
+fn load_snapshot(load_cmd: LoadCmd) {
+    let data = fs::read(Path::new(&load_cmd.source_dir).join("snapshot.borsh"))
+        .expect("Failed reading snapshot.borsh");
+
+    let snapshot: SpeedyCheckpoint = SpeedyCheckpoint::try_from_slice(data.as_ref()).unwrap();
+
+    let home_dir = Path::new(&load_cmd.target_home);
+    fs::copy(
+        Path::new(&load_cmd.source_dir).join("genesis.json"),
+        Path::new(&load_cmd.target_home).join("genesis.json"),
+    )
+    .unwrap();
+    fs::copy(
+        Path::new(&load_cmd.source_dir).join("config.json"),
+        Path::new(&load_cmd.target_home).join("config.json"),
+    )
+    .unwrap();
+
+    let config = nearcore::config::load_config(&home_dir, GenesisValidationMode::UnsafeFast)
+        .unwrap_or_else(|e| panic!("Error loading config: {:#}", e));
+    let store = init_and_migrate_store(home_dir, &config).unwrap();
+    let chain_genesis = ChainGenesis::from(&config.genesis);
+    let runtime = Arc::new(NightshadeRuntime::with_config(
+        home_dir,
+        store.clone(),
+        &config,
+        config.client_config.trie_viewer_state_size_limit,
+        config.client_config.max_gas_burnt_view,
+    ));
+    // This will initialize the database (add genesis block etc)
+    let _chain = Chain::new(
+        runtime.clone(),
+        &chain_genesis,
+        DoomslugThresholdMode::TwoThirds,
+        !config.client_config.archive,
+    )
+    .unwrap();
+
+    let mut store_update = store.store_update();
+    // Store epoch information.
+    write_epoch_checkpoint(&mut store_update, &snapshot.current_epoch);
+    write_epoch_checkpoint(&mut store_update, &snapshot.prev_epoch);
+    write_epoch_checkpoint(&mut store_update, &snapshot.next_epoch);
+
+    // Store blocks.
+    write_block_checkpoint(&mut store_update, &snapshot.block);
+    write_block_checkpoint(&mut store_update, &snapshot.prev_block);
+    write_block_checkpoint(&mut store_update, &snapshot.final_block);
+    write_block_checkpoint(&mut store_update, &snapshot.first_block);
```

Comment:
We need to check that snapshot is valid, that is, block must be the last block in prev_epoch, and it must be at least two epochs away from the current head of the chain. Both requirements seem hard to verify here in this binary, so I'm thinking, maybe we can write another python script that uses json rpc to query blocks info and verify that the snapshot is valid. There we can even do more verification, such as the block is actually on the chain.

The problem of that is, of course, then speedy-sync has to depend on the rpc endpoints provided by us, which is a kind of centralization. I think it is fine for now, since the current way of people starting nodes depend on our backup images, which is also centralized by us. And we do plan to implement EpochSync soon, which will be the ultimate decentralized way to syncing. 

I think this also solves the safety issue of speedy-sync, because the validity of checkpoint snapshot is guaranteed by our rpc. Otherwise, the user will run speedy-sync, and at the end, they still have to verify that their node is syncing to the current chain. Where do they verify the current chain, probably explorer, still our service. Verifying the checkpoint automatically is more user friendly so they don't have to worry about manually checking the chain later.

Reply:
The python script can be done in another PR, of course.